### ISSUE TITLE
Get Travis CI to Pass

### ISSUE THREAD

**Issue by acviana** [STScI-Citizen-Science/MTPipeline#102]
I'm creating a branch to get the tests organized and to get the Travis CI build to pass.

**Comment by acviana** [comment]
This build failed with the following status:

```
gcc -pthread -fno-strict-aliasing -DNDEBUG -g -fwrapv -O2 -Wall -Wstrict-prototypes -fPIC -I/sw/lib -I/home/travis/virtualenv/python2.7_with_system_site_packages/local/lib/python2.7/site-packages/numpy/core/include/numpy/ -I/usr/include/python2.7 -c src/lacosmicx.cpp -o build/temp.linux-x86_64-2.7/src/lacosmicx.o -O3 -funroll-loops

cc1plus: warning: command line option ‘-Wstrict-prototypes’ is valid for Ada/C/ObjC but not for C++ [enabled by default]

src/lacosmicx.cpp:15:19: fatal error: fitsio.h: No such file or directory

compilation terminated.

error: command 'gcc' failed with exit status 1

The command "python setup.py develop" failed and exited with 1 during .

Your build has been stopped.
```

Any test that imports from `imaging_pipeline.py` will therefore fail unless we can figure out how to build `fitsio.h` in Travis.

**Comment by acviana** [comment]
After the update in the last commit this build failed:

```
src/lacosmicx.cpp:15:19: fatal error: fitsio.h: No such file or directory
```

I think the issue is that the follow paths are hard-coded into the `setup.py` file:

```
cfitslib_path = '/sw/include'
cfitsinc_path = '/sw/lib'
```

So I'm going to add this to the `.travis.yml` file to try and work with that:

```
    - mkdir /sw
    ...
    - ./configure -prefix=/sw
```

**Comment by acviana** [comment]
So the `sudo` command _does_ work:

``` shell
$ sudo make install
/bin/cp -a libcfitsio* /sw/lib
/bin/cp fitsio.h fitsio2.h longnam.h drvrsmem.h /sw/include
/bin/cp cfitsio.pc /sw/lib/pkgconfig
```

But we still get the same error:

```
src/lacosmicx.cpp:15:19: fatal error: fitsio.h: No such file or directory
```

**Comment by ktfhale** [comment]
As of my [last commit](https://github.com/STScI-Citizen-Science/MTPipeline/commit/ec1012282ca4a84ef3725a654dbc5718938546d3) the pipeline build gets much farther. The setup script `setup.py` seems to crash around 7 minutes, after it has finished installing all dependencies for MTPipeline. It gets stuck at some kind of python exit function.

I had to manually set the C++ include and library paths. I'm not sure why this is necessary here, and not on our machines.

**Comment by ktfhale** [comment]
Tests now appear to be running on Travis, having also required it to install `astropy` and `drizzlepac` in the `setup.py` file. I think the first one fails:

```
F
======================================================================
FAIL: Iteratively generating tests.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/virtualenv/python2.7_with_system_site_packages/local/lib/python2.7/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/home/travis/build/STScI-Citizen-Science/MTPipeline/tests/unit_tests/test_imaging_pipeline.py", line 67, in check_make_output_file_dict
    assert result == expected_output, err_message
AssertionError: expected {'cr_reject_output': ['u2eu0101t_c0m.fits', 'u2eu0101t_cr_c0m.fits'], 'drizzle_weight': ['u2eu0101t_c0m_wide_single_wht.fits', 'u2eu0101t_c0m_center_single_wht.fits', 'u2eu0101t_cr_c0m_wide_single_wht.fits', 'u2eu0101t_cr_c0m_center_single_wht.fits'], 'drizzle_output': ['u2eu0101t_c0m_wide_single_sci.fits', 'u2eu0101t_c0m_center_single_sci.fits', .................
```

Could this be just because it doesn't have the test files?

**Comment by acviana** [comment]
:clap: :clap: :clap: :clap: 

Yay, tests. OK, now we'll just need to slowly reintegrate/redesign the tests. The issue is that I was originally writing end-to-end tests for the pipeline which require FITS files, databases, and heavily "test" the I/O of the code. I have since tried to separate this kind of testing, which is more like _"does this run?"_ from unit tests, which test the logic of the code. You can think of unit tests as _"does this do what we expect?"_. 

What I want is for the unit tests to run on Travis, because these are independent from the files. And then save the end-to-end tests to run locally.

**Comment by ktfhale** [comment]
So, after my last commit updating the the first test, Travis no longer even _gets_ to the tests. Needless to say, I am flummoxed, especially since the only change I made in the commit that failed building was to the python test files. I've been diffing the logs trying to see where it went wrong. Here are the most significant differences. The left log is from the successful build. The right log is from the failed build.

![screen shot 2014-07-03 at 10 36 09 am](https://cloud.githubusercontent.com/assets/7748162/3471153/79d863f6-02c0-11e4-8ff3-b301ccb50c36.png)

Looks like we're getting a new version of `Pillow`, which is apparently some kind of python image processing library. The build that fails is unable to find all of these image files, and a README. I don't know what to make of that, yet. Looks like Pillow 2.5.0 was released on GitHub two days ago.

![screen shot 2014-07-03 at 10 39 31 am](https://cloud.githubusercontent.com/assets/7748162/3471154/79d891c8-02c0-11e4-9a29-a6f7737311e1.png)

This is where the crash actually happens. Not much information, just "Exception in thread 14" right after `matplotlib` starts building. The crashed version apparently starts using 32 processors to do so, for some reason, and then one of those threads does something not good.

... I think I'm just going to try recomitting, to see if anything changes.

**Comment by ktfhale** [comment]
Unsurprisingly, running it again changed nothing.

It's clear that something is going wrong with the whole "Building using 32 processes" thing. The error that the built ultimately hangs with is:

```
Traceback (most recent call last):
  File "/usr/lib/python2.7/threading.py", line 551, in __bootstrap_inner                          
    self.run()                                                                                    
  File "/usr/lib/python2.7/threading.py", line 504, in run                                        
    self.__target(*self.__args, **self.__kwargs)                                                  
  File "/usr/lib/python2.7/multiprocessing/pool.py", line 319, in _handle_tasks                   
    put(task)                                                                                     
PicklingError: Can't pickle <type 'function'>: attribute lookup __builtin__.function failed       
No output has been received in the last 10 minutes, this potentially indicates a stalled build or 
The build has been terminated                                                                     
```

Something about pickling goes wrong. Pickling is python's object serializer. A serializer's apparently responsible for packing up and sending a function, object, or whatever to each process, so that process can work on it. It's an essential part part of parallel processing. [I'm just reading this here](http://matthewrocklin.com/blog/work/2013/12/05/Parallelism-and-Serialization/). But python's pickler can't handle functions, I'm reading, so the multiprocessing effort fails when python tries to package and send something to a process which can't be packaged.

Build 28, the build that actually completes and fails on the tests, never does any multiprocessing, so far as I can tell. It never tries "Building using 32 processes." I still don't know why Travis is suddenly trying to build python packages with multiple processes. It apparently does so successfully for `Pillow`, but crashes and burns with `Matplotlib`.

I haven't been able to find much online to help. There's a lot for this kind of pickling error, but I've found nothing yet that's related to this error and building packages. At the moment, I have a few ideas. First, roll back the branch to the commit for build 28, and see if that still works. If it doesn't, it's something that's changed about the Travis platform, or some of the third party packages we're installing. Second, find out why Travis is trying to do parallel processing stuff, and get it to stop.

Apparently you can just get Travis to run an old build again! No branch sorcery needed. Waiting for it to start...

**Comment by ktfhale** [comment]
Build 28, which previously built just fine, now fails, with the same problems as in build 30 and 32. So whatever's gone wrong with the build, I don't think it was us.

This makes me suspect Pillow, which was updated from 2.4.0 to 2.5.0 between builds 28 and 30. Building pillow is the first time we see the use of multiple processes in building, so I'm going to see if I can force us to use version 2.4.0. It seems unlikely that an image processing library would be the cause of all this, though.

EDIT: Before I pushed anything to the branch, I ran the setup.py script on my local copy with it set to use only version 2.5.0. I got the same errors Pillow gets in Branch 30 (no files found matching various image formats), and it reported "Building using 2 processes." If I don't force it to use version 2.5.0, setuptools on my mac is quite happy to use the 2.4.0 version sitiing in packages.

I'm now going to force the Travis build to use version 2.4.0.

**Comment by ktfhale** [comment]
[The build was successful, and all tests pass.](https://travis-ci.org/STScI-Citizen-Science/MTPipeline/builds/29090476)

Forcing it to use the 2.4.0 version of Pillow should have been the first thing I tried. Why installing Pillow results in changing how Matplotlib is built, I have no idea. But, things seem to work now.

Should `travis-fix` be merged with `master`, after `travis` is? Also, should we possibly send something to the Pillow people?

**Comment by acviana** [comment]
![screen shot 2014-07-06 at 8 50 25 pm](https://cloud.githubusercontent.com/assets/1630700/3490537/ca6853dc-0570-11e4-9bf0-d75deca26e47.png)

This is the most beautiful things I've ever seen.

Also, you should update setup.py to force the same version of Pillow. If you feel you can isolate the issue it would be great top open a PR with the Pillow Team.

**Comment by ktfhale** [comment]
I rebased my local copy of `travis-fix` with `master`, then merged my local copy of `travis-fix` with the remote repository. After doing so, `travis-fix` no longer successfully builds on Travis. I'm rerunning the previously successful build#33 to see whether or not it's something we changed.

It appears to fail when doing something with `astropy` involving Cython. This is concerning, because I manually set the `CPLUS_INCLUDE_PATH` and `LIBRARY_PATH`, so maybe that's screwing something up.

```
error: Setup script exited with error: astropy/time/erfa_time.c: Could not
find C file astropy/time/erfa_time.c for Cython file astropy/time/erfa_time.pyx
when building extension astropy.time.erfa_time. Cython must be installed to
build from a git checkout.
```

Did the dependencies in `master` change recently? I didn't note any conflicts in `setup.py` while rebasing, however, so I wouldn't expect it.

**Comment by ktfhale** [comment]
Bizarrely, build#33, which previously built and ran just fine, now still appears to build and run its tests successfully (we get exit code 0), but Travis's indicator button reports it as errored... and it doesn't seem to be running for long enough, so something's definitely fishy.

**Comment by ktfhale** [comment]
It seems like my browser's version of the log simply wasn't refreshing when I reloaded. The actual text file shows the same `astropy`/Cython error for rerunning build#33 as in build#35. Build#33 was successful on Friday, so something has changed in the meantime.

When build#33 ran successfully, it used `astropy` version 0.3.2. Now, when rerun, build#33 uses `astropy` version 0.4rc1, which was released six days ago on GitHub. The build after the rebase, build#35, also uses this newer version. I'm going to try forcing version 0.3.2.

EDIT: the new version 0.4 of `astropy` mentions changes to the how the cython compiler is handled. It references Ticket#1198, https://github.com/astropy/astropy/pull/1198, which deals explicitly with issues on Travis.
